ssl/statem/statem_clnt.c

```diff
@@ -1377,7 +1378,7 @@ MSG_PROCESS_RETURN dtls_process_hello_verify(SSL_CONNECTION *s, PACKET *pkt)
      */
     if ((max_version == DTLS1_3_VERSION || max_version == 0)
             && !SSL_set_max_proto_version(ssl, DTLS1_2_VERSION)) {
-        SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_LENGTH_MISMATCH);
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
```

Comment:
Is this really an internal error here? We had min version set at 1.3 - could any correct server respond with HelloVerify? I.e. would SSL_AD_UNEXPECTED_MESSAGE be more correct here?

Reply:
Hmm. Didn't think of that. I guess this would fail for a DTLS 1.3 only client connecting to a DTLS 1.2 (or less) server. So maybe 
```
        SSLfatal(s, SSL_AD_PROTOCOL_VERSION, SSL_R_UNSUPPORTED_PROTOCOL);
```
Is this most appropriate?

Reply:
If min_version is DTLSv1.3 then `SSL_AD_PROTOCOL_VERSION` seems correct since it implies no DTLSv1.3

However, I don't think it is the correct approach to call `SSL_set_max_proto_version` here. This is a user API for setting the max allowed protocol version. It shouldn't be (mis)used for constraining what is actually selected during the handshake.

Rather I would somehow detect this in `ssl_choose_client_version` (probably setting `real_max` if we detect a hello verify request has been received...we probably need some additional field to remember this)

Reply:
> Rather I would somehow detect this in ssl_choose_client_version

Looked at it. Before this fix we never get as far as `ssl_choose_client_version` because that gets called in `tls_process_server_hello` and the assertion is hit while preparing the second DTLSv1.3 ClientHello. Ie. before the ServerHello is sent.

I'll continue digging.

Reply:
Agree and I've implemented that part. But I haven't (yet) found a way to construct the second ClientHello as a non-DTLSv1.3 client. Maybe you can point me in the right direction?

Reply:
Shouldn't we basically be creating an identical ClientHello with the exception of the cookie? So, if we went this route then the second ClientHello would still have the supported versions extension indication DTLSv1.3 support (even though actually selecting DTLSv1.3 would be a protocol violation).

The RFCs seems a little unclear on how exactly a DTLSv1.2/DTLSv1.3 dual stack implementation should process a HelloVerifyRequest. The DTLSv1.3 RFC says very little about it other than:

> A dual-stack DTLS 1.2 / DTLS 1.3 client MUST, however, be prepared to interact with a DTLS 1.2 server.

The DTLSv1.2 RFC says this:

> When responding to a HelloVerifyRequest, the client MUST use the same
>    parameter values (version, random, session_id, cipher_suites,
>    compression_method) as it did in the original ClientHello.

This doesn't actually explicitly mention the extensions at all, but it is kind of implied by the fact it "MUST" us the same parameter values.

So my (possibly wrong) interpretation is that the 2nd ClientHello should be the same as the first with the exception of the added cookie.

Reply:
Yes you are right. So handling it should be more or less like hrr. Which is also what the assertion hints at.
Thanks! I think that was the missing piece.